### ISSUE TITLE
multiple popups "has been modified locally, overwrite object?"

### ISSUE THREAD

**Issue by larshp** [larshp/abapGit#515]
show list instead

**Comment by larshp** [comment]
@sbcgua I would like to create a new page to replace the multiple popups, something like

```abap
CREATE OBJECT lo_select.
lcl_app=>gui( )->call_popup( lo_select ).
lt_selected = lo_select->get_selected( ).
```

instead of calling

```abap
lv_answer = lcl_popups=>popup_to_confirm(
  titlebar              = 'Warning'
  text_question         = lv_question
  display_cancel_button = abap_false ).
```

multiple times from INCLUDE zabapgit_objects_impl, note that this include is defined before the GUI classes.

Any thoughts on how to do this? The GUI classes has to be defined before the application logic, but currently gui uses lcl_gui_router, which references the application logic. Perhaps replace lcl_gui_router with a interface reference in lcl_gui?

**Comment by sbcgua** [comment]
Good task for the new year :)

Well, yes, I had some thoughts about Pull page (let's call it like this). I think we should look at it a bit wider that just a confirmation dialog. I cannot check the code now but I think all this warning stuff should be a separate pre-step to the actual pull - it is a good moment to split this logic from deserialize. (Alternatively it should be an ALV dialog, not html page)

E.g. user presses pull (or switch branch), this issues a `go_pull` event (a route) that can be processed normally in the current concept - show a page like stage with all the needed info and **interaction** (there might be more than just confirmations). Then user confirms Pull and the page starts actual deserialize routine. (Deserialize can theoretically double check overwriting again but then issue simple error - kind of "system changed before you initiated pull and confirmed it" - which definitely very rare).

Also ! I was thinking that the Pull page can actually check more than just overwrites. E.g. it can check any inactive objects, and if anyone blocks any code to be deserialized - this is currently not checked and some times leads to really undertermined system state which is quite annoying. Maybe some more warnings, maybe some cross-abap-version deserialization options (????).

Also, the page can allow some partial individual pulls or some kind of cherry-picking which was asked by @nununo  some time ago somewhere in issues.

Well, these are my thoughts. :) Happy new year by the way ! :))

**Comment by larshp** [comment]
yeah, one step at a time :smile: All are good ideas, but i guess we need pull/popup pages(guess we need to find a better term, as "pull" is also a git term) in all cases.

I would like to do a quick fix using abapGit pages instead of an ALV. We can have this implemented quite fast and deliver value to the users. The excellent ideas above will take longer

**Comment by sbcgua** [comment]
"quick fix" ... mmmm ... I'm a bit skeptical :) But yeah, there should be something simple to start from.

So for the first release I propose then the following flow: 
- user runs pull or branch switch (is that all ??)
- UI inits the pull page
- pull page calls regular sync routines which result in the  list of objects to update
- Pull page lists all objects to update. The "dirty" ones are marked in color + have check box or some "skip" button whatever which allow to EXCLUDE the object from deserialization.
- then there is a button "Confirm" which actually continues deserialization

Does it look logical ?

Unclear:
- Should it show the page if there are no conflicts ? 
- What worries me is this partial desealization - it is a very strange state imho - was the pull/switch complete if I excluded an object ? Shell we update the commit hash ? This operation is not atomic in some way. I think this concept should be reconsidered somehow on the way.

**Comment by nununo** [comment]
I was about to submit this exact issue :) but first I looked for a similar one and here it is! you guys are great! Thanks!

**Comment by larshp** [comment]
lets try to move it forward then :o)

Above plan looks good, but there are a lot of questions that need answers, and testing to be done.

So I would still like to add the possibility for a popup/pull page, this we can deliver quite fast, and it will deliver value to the users. I guess popup/pull pages are still needed in the overall abapGit UI framework anyhow.

**Comment by larshp** [comment]
with https://github.com/larshp/abapGit/pull/559 being merged, this issue is becoming more important

**Comment by larshp** [comment]
currently we have the following checks:
- master language
- check_requirements (ZABAPGIT_REQUIREMENTS)
- overwrite locally changed object
- overwrite from other package
- compare_remote_to_local (db changes)

refactor so the checks can be run decoupled from the deserialization, and in background or foreground

**Comment by EduardoCopat** [comment]
Fixed by #934?

**Comment by larshp** [comment]
yeah, closing
